\documentclass[10pt]{article} %Sets the default text size to 11pt and class to article.
%------------------------Dimensions--------------------------------------------
\topmargin=0.0in %length of margin at the top of the page (1 inch added by default)
\oddsidemargin=0.0in %length of margin on sides for odd pages
\evensidemargin=0in %length of margin on sides for even pages
\textwidth=6.5in %How wide you want your text to be
\marginparwidth=0.5in
\headheight=0pt %1in margins at top and bottom (1 inch is added to this value by default)
\headsep=0pt %Increase to increase white space in between headers and the top of the page
\textheight=9.1in %How tall the text body is allowed to be on each page

\usepackage{hyperref}
\usepackage[usenames,dvipsnames]{color}
\usepackage{paralist}

\newcounter{pubCtr}

\begin{document}

\centerline {\large\bf Frank Austin Nothaft} 
%\centerline {}
\centerline {\href{mailto:fnothaft@alumni.stanford.edu}{fnothaft@alumni.stanford.edu} $\bullet$ 202.340.0466}
\noindent {}

\begin {minipage}[t]{0.2\linewidth}
\vspace{0pt}
\noindent {\bf Work \\ Experience}
\end {minipage}
\begin {minipage}[t]{0.8\linewidth}
\vspace{0pt}
\noindent {{\bf Tempus Labs}} \\
\centerline {{\bf Senior Director, Platform Services Engineering \hfill November 2020--present}}
\begin{itemize}
\item {I lead the engineering and computational biology teams supporting the development of the \href{https://www.tempus.com/life-sciences/lens/}{Tempus Lens} platform, a software-as-a-service solution for oncology research. I have led multiple product launches across a broad range of platform components, including a web application for rapid biostatistical (e.g., rwOS) and molecular (e.g., co-mutation landscape) analysis of large multimodal oncology datasets, a cloud-hosted data science environment for computational biology analyses, and multiple new de-identification implementations.}
\item {As a Senior Director accountable for the commercial success of the platform, I established both a pre-sales computational biology team and customer success/documentation team to support our strategic collaborations. I work closely with scientific and IT stakeholders at our research collaborators and within Tempus' professional science services teams to drive adoption of Lens. Internally, I work with key functions (site reliability engineering, finance/accounting) to drive improvements to core KPIs, including a multi-million dollar reduction in cloud costs, 10$\times$ reduction in  latency through key product endpoints, and improvement of data science platform uptime SLI.}
\item {As a hiring manager across multiple functions, I have created a hiring process that allowed our teams to scale from 10 to 40 employees while maintaining a high talent bar in both engineering and science functions. This hiring process also allowed our team to improve both gender and URM-status diversity by 20\%, at both the IC and managerial level.}
\end{itemize}

\noindent {{\bf Databricks }} \\
\centerline {{\bf Technical Director, Health/Life Sciences \hfill January 2019--November 2020}}
\centerline {{\bf Go-To-Market Lead, Genomics \hfill January 2018--December 2018}}
\centerline {{\bf Genomics SME (Consultant) \hfill February 2017--December 2017}}
\begin{itemize}
\item {I led the development of the Databricks Runtime for Genomics from initial conception through general availability, across a team of three engineers and two bioinformaticians dedicated to the product. We delivered a high performance implementation of the GATK4 best practices germline, somatic, and joint genotyping pipelines, delivered as a managed cloud service. We achieved a 3$\times$ latency improvement over the Edico DRAGEN platform with comparable cost and accuracy.}
\item {I supported multiple key pharmaceutical partners who were part of the UK Biobank exome sequencing effort. I led the development of a
strategic product partnership with the Regeneron Genetics Center around Project Glow, an Apache Spark based tool for managing large genomic variation datasets that provided a $10\times$ improvement in
performance for genome wide association studies at the million patient scale.}
\item {In the latter half of my tenure, I also assumed responsibility for all go-to-market activities (marketing, sales, partnerships/business development, and
field engineering) across our healthcare/life sciences customers. From 2018 through the end of the first half of 2020, we grew
committed customers by $7\times×$ and annual recurring revenue by $12\times$. Our clients
included eight of the top ten global pharmaceutical companies, four of the top five
US payers, and three of the top ten US hospitals.}
\end{itemize}

\end {minipage}

\begin {minipage}[t]{0.2\linewidth}
\vspace{0pt}
\noindent {\bf Work \\ Experience \\ (continued)}
\end {minipage}
\begin {minipage}[t]{0.8\linewidth}
\vspace{0pt}

\noindent {{\bf UC Berkeley AMPLab }} \\
\centerline {{\bf Research Assistant \hfill August 2013--December 2017}}
\begin{itemize}
\item {I led development of the ADAM/Big Data
Genomics tool suite which provided an open source, Apache Spark-based implementation of the ``GATK Best Practices'' pipeline. Our pipeline could process a 30$\times$ WGS in less than 1 hour, with comparable accuracy to the GATK, and grew to have more than 75 open source contributors.}
\item {In collaboration with UCSC, I contributed to the Toil
workflow runner and worked to compare the performance of Kalisto vs. STAR/RSEM RNA-seq quantification pipelines through a joint analysis of more than 20,000 RNA-
seq samples, culminating in a publication in Nature Biotechnology.}
\item {As the primary grantwriter for the AMPLab Genomics team, I led cross-institute
proposal efforts for successful NIH BD2K (with UCSC, OHSU, Sage, and UCSF),
Cancer Cloud Pilot (with UCSC and Broad Institute), and SBIR grants (with
Curoverse), as well as a Chan Zuckerberg Institute grant.}
\end{itemize}

\noindent {{\bf Color Genomics }} \\
\centerline {{\bf Bioinformatics Consultant \hfill October 2013--December 2014}}
\noindent {Developed a germline copy number variant calling pipeline that ran in a CAP/CLIA validated setting as part of a targeted NGS panel for hereditary cancer screening.} \\

\centerline {{\bf Broadcom, R\&D Engineer, IC Design 2 \hfill March 2016--June 2016}}
\centerline {{\bf Broadcom, Engineer, Staff 1---IC Design \hfill April 2012--March 2016}}
\centerline {{\bf Broadcom, Engineer---IC Design \hfill June 2011--April 2012}}
\noindent {I developed novel validation techniques for cellular, BlueTooth, WiFi, and
802.11ad chipsets. My techniques enabled modeling circuit dynamics, and hardware
accelerated circuit simulation.} \\

\end {minipage}

\begin {minipage}[t]{0.2\linewidth}
\vspace{0pt}
\noindent {\bf Education}
\end {minipage}
\begin {minipage}[t]{0.8\linewidth}
\vspace{0pt}
\noindent {{\bf University of California, Berkeley}} \\
\noindent {Doctor of Philosophy, Computer Science. August 2013--December 2017. GPA: 3.81.} \\
\noindent {Masters of Science, Computer Science. August 2013--May 2015. GPA: 3.79.} \\
\noindent { Advisors: Dave Patterson and Anthony Joseph.} \\

\noindent {{\bf Stanford University}} \\
\noindent {Bachelor of Science with Honors, Electrical Engineering. September 2007--June 2011.} \\
\noindent {Minor in Management Science $\&$ Engineering. GPA: 3.24.} \\
\end {minipage}

\begin {minipage}[t]{0.2\linewidth}
\vspace{0pt}
\noindent {\bf Selected \\ Publications}
\end {minipage}
\begin {minipage}[t]{0.8\linewidth}
\vspace{0pt}

\noindent {Full publication list available at \url{https://scholar.google.com/citations?user=
BYSy-9oAAAAJ}.}

\setdefaultleftmargin{13pt}{}{}{}{}{}
\begin{itemize}
\item {Lisa Wu, David Bruns-Smith, \textbf{Frank Austin Nothaft}, Qijing Huang, Sagar
Karandikar, Johnny Le, Andrew Lin, Howard Mao, Brendan Sweeney, Krste Asanovi\'{c}, David A Patterson, Anthony D Joseph. “FPGA Accelerated INDEL Re-alignment in the Cloud.” In Proceedings of the International Symposium on High
Performance Computer Architecture, February 2019 (HPCA ’19).}
\item {Alyssa Kramer Morrow, George Zhixuan He, \textbf{Frank Austin Nothaft}, Justin Paschall,
Nir Yosef, Anthony D. Joseph. ``Mango: Distributed Visualization for Genomic Analysis.'' In Cell Systems, December 2019. Originally posted as
\emph{BioR$\chi$iv:360842}, July 2018.}
\end{itemize}
\end {minipage}

\begin {minipage}[t]{0.2\linewidth}
\vspace{0pt}
\noindent {\bf Selected \\ Publications \\ (continued)}
\end {minipage}
\begin {minipage}[t]{0.8\linewidth}
\vspace{0pt}

\begin{itemize}
\item {Michael D. Linderman, Davin Chia, Forrest Wallace, and \textbf{Frank Austin Nothaft}, ``DECA:
Scalable XHMM exome copy-number variant calling with ADAM and Apache Spark.'' In BMC Bioinformatics, December 2019. Originally posted as
\emph{BioR$\chi$iv:062497}, September 2017.}
\item {John Vivian, Arjun Rao, \textbf{Frank Austin Nothaft}, Christopher Ketchum,
Joel Armstrong, Adam Novak, Jacob Pfeil, Jake Narkizian, Alden D. Deran,
Audrey Musselman-Brown, Hannes Schmidt, Peter Amstutz, Brian Craft, Mary Goldman,
Kate Rosenbloom, Melissa Cline, Brian O'Connor, Megan Hanna, Chet Birger,
W. James Kent, David A. Patterson, Anthony D. Joseph, Jingchun Zhu,
Sasha Zaranek, Gad Getz, David Haussler, and Benedict Paten. ``Toil enables reproducible,
open source, big biomedical data analyses.'' In \emph{Nature
Biotechnology}, April 2017. Originally posted as \emph{BioR$\chi$iv:062497}, July 7, 2016.}
\item {\textbf{Frank Austin Nothaft}, Matt Massie, Timothy Danford, Zhao Zhang,
Uri Laserson, Carl Yeksigian, Jey Kottalam, Arun Ahuja, Jeff Hammerbacher,
Michael Linderman, Michael J. Franklin, Anthony D Joseph, and David A. Patterson.
``Rethinking data-intensive science using scalable analytics systems.'' In
\emph{Proceedings of the International Conference on Management of Data},
May~2015 (SIGMOD~'15).}
\item {Selected industrial white papers:}
\begin{itemize}
\item {\href{https://www.databricks.com/blog/2020/06/25/introducing-glowgr-an-industrial-scale-ultra-fast-and-sensitive-method-for-genetic-association-studies.html}{Introducing GlowGR: An industrial-scale, ultra-fast and sensitive method for genetic association studies}. \emph{Joint white paper with the Regeneron Genetics Center discussing the sensitivity of the GlowGR GWAS method.} June 2020.}
\item {\href{https://www.databricks.com/blog/2019/06/19/accurately-building-genomic-cohorts-at-scale-with-delta-lake-and-spark-sql.html}{Accurately Building Genomic Cohorts at Scale with Delta Lake and Spark SQL}. \emph{Analysis of the accuracy and scalability of the Databricks joint genotyping pipeline.} June 2019.}
\item {\href{https://www.databricks.com/blog/2018/09/10/building-the-fastest-dnaseq-pipeline-at-scale.html}{Building the Fastest DNASeq Pipeline at Scale}. \emph{Performance benchmarking of Databricks Genomics pipeline vs. Edico DRAGEN}. September 2018.}
\end{itemize}
\end{itemize}
  
\noindent{}
\end {minipage}

\end{document}
